src/backends/sqlite3/session.cpp

```diff
@@ -328,3 +386,12 @@ sqlite3_blob_backend * sqlite3_session_backend::make_blob_backend()
 {
     return new sqlite3_blob_backend(*this);
 }
+
+details::prepare_temp_type prepare_column_descriptions_statement(
+        details::prepare_type& prepare, std::string & table_name, soci::column_info& info)
+{
+    // trick the compiler to use the static type sqlite_column_info instead of the standard one
```

Comment:
This cast is undefined behaviour and really shouldn't be done. With the current API the obvious alternative is to create a local `sqlite_column_info` object and copy it to `info` later, but it would be great if we could have a nicer API not requiring passing `info` here at all.

Reply:
Will change it to `reinterpret_cast<sqlite_&>` instead, which is not ub.

Reply:
It's still UB because the static type doesn't match the cast target.

But the problem is not "just" the UB, it's really problematic to perform a cast to a wrong type.

Reply:
Definitely not an UB, the case is described in cppreference : https://en.cppreference.com/w/cpp/language/reinterpret_cast#Notes

Reply:
I will correct the code this week-end and propose others implementations but they will likely be worse and introduce larger API changes.
In a static context the only solution I have in mind is :
* Calling another function (what I am doing with the reinterpret cast)
* Having dynamic type in the parameters that are aware of the backend :
  * By having a field in the column_info (or another type)
  * By making the colum_info virtual (or another type)
 

In the client, it will be something like this :

```c++
namespace soci {
  struct dynamic_column_info : public column_info {
    virtual void from_base(values const & v, indicator ind) {
      soci::type_conversion<column_info>::from_base(v, ind, &this);
    }
  
  template<>
  struct type_conversion<dynamic_column_info> {
    typedef values base_type;
    static void from_base(values const & v, indicator ind, dynamic_column_info & ci) {
      ci.from_base(v, ind);
    }
  } /* struct tc */
} /* !namespace soci */

std::string tableName;
std::unique_ptr<dynamic_column_info> ci = sql.create_backend_typed_column_info();
soci::statement stmt = (sql.prepare << sql.prepare_column_descriptions(tableName), soci::into(*ci));
```

I admit, it will be cleaner in the backend but in the user code, not so much.

Reply:
Sorry, where do you see anything allowing to cast a pointer to an object to a pointer of a type different from the actual run-time type in these notes?

Reply:
```c++
struct S { int x; };
struct S1 : S {};    // standard-layout
 
S1 s1 = {};
auto p1 = reinterpret_cast<S*>(&s1); // value of p1 is "pointer to the S subobject of s1"
auto i = p1->x; // OK
p1->x = 1;      // OK
```

Reply:
This cast is in another direction. Of course you can cast a pointer to the derived class to the pointer of a base class. But you're doing it the other way round and when the object being pointed to is _not_ an object of the derived class.